### PR TITLE
Fixes default value of `softmax_scale` in `PhiFlashAttention2`.

### LINKED ISSUE #28488: fine tuning the updated Phi-2 with flash-attn-2 produces very high loss > 2
### System Info

The updated code of phi-2 produces a high loss, I have tried fp16, bf16, deepspeed and fsdp the result is the same -> loss starts at 2 and keeps going higher. Setting `use_flash_attention_2=False` fixes this or using the old phi-2 modeling file.

torch==2.1.2
flash-attn==2.4.2
transformers==4.37.0.dev0

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Fine-tune the updated phi-2 model using transformers trainer

### Expected behavior

Loss go down

### PR DESCRIPTION
**Pull request by gugarosa**
# What does this PR do?

- Phi has never used `softmax_scale=1.0` with Flash-Attention, so the default is being moved to `None`. This tentatively fixes any issue regarding fine-tuning Phi-based checkpoints when Flash-Attention 2 is turned on.

- Documentation is also updated to reflect the official Phi checkpoints.

Fixes #28488 (tentative)

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @susnato

### CODE REVIEW

File: docs/source/en/model_doc/phi.md [hunk 1]
```diff
@@ -60,32 +60,32 @@ including hallucinations and the potential for toxic and biased generations –e
...
 
 </Tip>
 
 ```python
 >>> from transformers import AutoModelForCausalLM, AutoTokenizer
 
->>> model = AutoModelForCausalLM.from_pretrained("susnato/phi-2")
->>> tokenizer = AutoTokenizer.from_pretrained("susnato/phi-2")
+>>> model = AutoModelForCausalLM.from_pretrained("phi-2")
+>>> tokenizer = AutoTokenizer.from_pretrained("phi-2")
```
Comment (susnato) [inline docs/source/en/model_doc/phi.md:88]: Thanks for changing these checkpoints. 🙌 
I was about to open a PR to change these

File: src/transformers/models/phi/modeling_phi.py [hunk 2]
```diff
@@ -506,7 +506,7 @@ def forward(
             value_states = value_states.to(target_dtype)
 
         attn_output = self._flash_attention_forward(
-            query_states, key_states, value_states, attention_mask, q_len, dropout=attn_dropout, softmax_scale=1.0
+            query_states, key_states, value_states, attention_mask, q_len, dropout=attn_dropout, softmax_scale=None
         )
```
Comment (susnato) [inline src/transformers/models/phi/modeling_phi.py:510]: Ah, was this the reason for the issue regarding fine-tuning? 
Now I am curious how the FA tests were passing before...

Anyway thanks a lot for fixing this!

**Comment by gugarosa** [comment]
The `loss=0.0` error while fine-tuning with FP16 is another issue and I do have an ugly fix, but will look into it with more patience (and use a separate PR).

File: src/transformers/models/phi/modeling_phi.py (hunk 2 above)
Comment (susnato) [reply src/transformers/models/phi/modeling_phi.py:510]: BTW just to be sure, could you please run all flash attention tests(for `phi`) to check if they are passing or not.

```
RUN_SLOW=1 pytest -m flash_attn_test tests/models/phi --verbose
```

Should not be necessary since it's already fixing the fine-tuning issue, but just to be sure.

Comment (gugarosa) [reply src/transformers/models/phi/modeling_phi.py:510]: Just ran and everything has passed!

Comment (ArthurZucker) [reply src/transformers/models/phi/modeling_phi.py:510]: Yep, our CIs don't test flash attention, bit of a pity!

**Review by ArthurZucker — APPROVED** [review]
Thanks a lot for this fix! Very tricky indeed

**Comment by gugarosa** [comment]
No problems! Thanks for the merge!

**Comment by younesbelkada** [comment]
Thanks very much @gugarosa for the deep dive and the fix!
